scripts/sensor/_04b_time_frequency_csp.py

```diff
@@ -0,0 +1,436 @@
+        )
+
+        X = epochs_filt.get_data()
+
+        # We apply PCA before running CSP:
+        # - much faster CSP processing
+        # - reduced risk of numerical instabilities.
+        X_pca = pca.fit_transform(X)
+        del X
+
+        cv_scores = cross_val_score(
+            estimator=clf,
+            X=X_pca,
+            y=y,
+            scoring=cfg.decoding_metric,
+            cv=cv,
+            n_jobs=1,
+        )
+
+        # The row is a view on the original dataframe, so we can modify it here
+        # and the dataframe will be updated.
+        row['mean_crossval_score'] = cv_scores.mean()
+
+    # Loop over times x frequencies
+    #
+    # Note: We don't support varying time ranges for different frequency
+    # ranges to avoid leaking of information.
+
+    time_bins = np.array(cfg.decoding_csp_times)
+    if time_bins.ndim == 1:
+        time_bins = np.array(
+            list(zip(time_bins[:-1], time_bins[1:]))
+        )
+    assert time_bins.ndim == 2
+
+    tf_decoding_table = pd.DataFrame(
+        columns=[
+            'subject', 'cond_1', 'cond_2', 't_min', 't_max', 'f_min',
+            'f_max', 'freq_range_name', 'mean_crossval_score', 'metric'
+        ]
+    )
+
+    for freq_range_name, freq_bins in freq_name_to_bins_map.items():
+        for time_bin in time_bins:
+            t_min, t_max = time_bin
+
+            for freq_bin in freq_bins:
+                f_min, f_max = freq_bin
+                one_row = pd.DataFrame(
+                    {
+                        'subject': subject,
+                        'cond_1': condition1,
+                        'cond_2': condition2,
+                        't_min': t_min,
+                        't_max': t_max,
+                        'f_min': f_min,
+                        'f_max': f_max,
+                        'freq_range_name': freq_range_name,
+                        'mean_crossval_score': np.nan,
+                        'metric': cfg.decoding_metric
+                    },
+                    index=[0]
+                )
+                tf_decoding_table = pd.concat(
+                    [tf_decoding_table, one_row],
+                    ignore_index=True
+                )
+
+    for _, row in tf_decoding_table.iterrows():
+        tmin = row['t_min']
+        tmax = row['t_max']
+        fmin = row['f_min']
+        fmax = row['f_max']
+        cond1 = row['cond_1']
+        cond2 = row['cond_2']
+        freq_range_name = row['freq_range_name']
+
+        msg = (
+            f'Contrast: {cond1} – {cond2}, '
+            f'Freqs (Hz): {fmin}–{fmax} '
+            f'({freq_range_name}), '
+            f'Times (s): {round(tmin, 3)}–{round(tmax, 3)}'
+        )
+        logger.info(
+            **gen_log_kwargs(msg, subject=subject, session=session)
+        )
+
+        epochs_filt, y = prepare_epochs_and_y(
+            epochs=epochs, contrast=contrast, fmin=fmin, fmax=fmax, cfg=cfg
+        )
```

Comment:
so here you filter the same way for each time bin so you seem to redo many time the same costly temporal filtering.